### PR TITLE
Use spanner via java api (instead of spring jdbc)

### PR DESCRIPTION
**Pull request by yardenvan**
* added asset table creation
* added all of the 6 assets info to be added to the tables
* removed spring (as it was only used for the jdbc which was removed)
* added class documentations

### CODE REVIEW

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 1]
```diff
@@ -1,49 +1,46 @@
...
 import java.util.Map;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-@Table(name = "Assets")
+/**
+ * The AssetObject class is an abstract class which is the parent of all of the specific asset
+ * object classes. This class provides setters and getters to all of the common fields of the
+ * asset objects. In addition it provides some helper functions for specific type parsing and
+ * conversions which are used in several asset object classes (such as a convertStringToDate function).
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java:21]: right justify (the IDE should be able to do this for you

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 2]
```diff
@@ -1,49 +1,46 @@
...
+ */
 abstract public class AssetObject {
-    @PrimaryKey(keyOrder = 1)
-    private String accountId = ProjectConfig.getInstance().getAccountId();
-    @PrimaryKey(keyOrder = 2)
-    private String projectId = ProjectConfig.getInstance().getProjectId();
-    // We are using both the asset name and kind in order to ensure uniqueness
-    @PrimaryKey(keyOrder = 4)
+    // asset primary keys
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: We usually write comments as full sentences, capitalized and period at the end
so "Asset primary keys"
I think the period is negotiable (it isn't really a sentence), but the capital is almost always there

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 3]
```diff
@@ -1,49 +1,46 @@
...
-    @PrimaryKey(keyOrder = 3)
-    @Column(name = "assetName")
     protected String name;
 
-    @Column(name = "assetId")
+    // asset additional data
     protected String id;
-    @Column(name = "assetType")
     protected String type;
     protected String zone;
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: Will this always be a zone, or can this also be a region, or even "global".
I'm wondering if we should generalize this as 'location'...?

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 4]
```diff
@@ -1,49 +1,46 @@
...
-    protected Date creationTime;
+    protected Timestamp creationTime;
     protected String status;
 
-    @NotMapped
+    protected AssetTypes assetTypeEnum;
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: Can you add a comment explaining the relationship between 'kind' and this field?

Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: Two naming comments:
1. I actually don't see the definition here, but I think it's more common to define enum type-names as singular "AssetType"
2. Unless there is a risk of confusing this with some other asset-type, I think this can simply be called "assetType" and the readers can find that it's an enum themselves. (unless I'm missing some other consideration)

Also many places below

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 5]
```diff
@@ -1,49 +1,46 @@
...
+
     private static final FluentLogger logger = FluentLogger.forEnclosingClass();
-    @NotMapped
     private static final Pattern LAST_SEGMENT_PATTERN = Pattern.compile(".*/");
-    @NotMapped
     private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
 
     protected abstract static class BaseBuilder<T extends AssetObject, B extends BaseBuilder> {
-        protected Map<String,String> assetObjectsMap;
+        protected Map<String,Object> assetObjectsMap;
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: What are the objects being held here?
If these are properties, I think the name 'assetPrpertiesMap' or simply 'assetProperties' will be more readable

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 6]
```diff
@@ -126,6 +131,22 @@ public T build() {
     }
 
     // AssetObject Class Getters
+    /**
+     * Get the accountId field of this object.
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: This is just me being a stickler, and there are many many examples of this rule not being followed, so feel free to ignore if this is too difficult or doesn't seem right, but the formal readability guidance is to use the indicative mood rather than the imperative mood under the understanding that we describe what the code does rather than tell it what to do. So "Returns the accountId..."

This is a general comment about style. For this case, I'd say that the method is actually self-descriptive enough to not require a comment at all 8-)

(also below)

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 7]
```diff
@@ -182,32 +203,61 @@ public String getStatus() {
         return this.status;
     }
 
+    /**
+     * Get the AssetType of this object.
+     * @return A enum of AssetTypes representing the specific type of this Asset Object.
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: "An enum"

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 8]
```diff
@@ -182,32 +203,61 @@ public String getStatus() {
...
     /*
      * This function receives a string representing a url and extracts the string after the last '/'
      * char in the url.
      * If the provided string does not match that pattern the original string is returned.
      */
-    protected static String getLastSeg(String url) {
-        Matcher matcher = LAST_SEGMENT_PATTERN.matcher(url);
+    protected static String getLastSeg(Object url) {
+        String urlToReturn = (String) url;
+        Matcher matcher = LAST_SEGMENT_PATTERN.matcher(urlToReturn);
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: IIUC 'matcher' should be "greedy" in order for this to work as described.
Are you sure it is?
Can you rephrase this in a way that will remove this ambiguity?

Comment (itaitz) [reply src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: also I think it would be more readable if it didn't require multiple matches.

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 9]
```diff
@@ -182,32 +203,61 @@ public String getStatus() {
...
         }
-        return url;
+        return urlToReturn;
+    }
+
+    /*
+     * This function receives a list of strings representing urls and returns a list of strings
+     * with the last segment for each url.
+     */
+    protected static List<String> convertListToLastSegList(Object urlsObject) {
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java:203]: I wonder if this can be rephrased as a ".stream().map(url ->" etc, possibly being so short as to allow to removing this function entirely

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java [hunk 10]
```diff
@@ -182,32 +203,61 @@ public String getStatus() {
...
         } catch (ParseException exception) {
             String error_msg = "Encountered a date parsing error. Dates should be in " +
                                 "yyyy-MM-ddTHH:mm:ss format, provided date: " + dateString;
             logger.atInfo().withCause(exception).log(error_msg);
         }
         return null;
     }
+
+    /*
+    This function receives a string representing an int and returns it as an int.
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: What happens if the types or formats don't match?

**Review by itaitz — COMMENTED** [review]
Looks very nice!
I only had a chance to look over part of this. Will continue tomorrow, but sending what I have until now

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java (hunk 4 above)
Comment (yardenvan) [reply src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: Regarding the second comment - in the asset table in our spanner DB the "type" column is called "assetType" and that was the reason for the more descriptive name "assetTypeEnum". The logic behind the "type" and "assetType" inconsistency was that when reading from the asset tables it might be important for the column names to have more meaningful names rather than the fields in the asset objects which are obviously asset related. Maybe this inconsistency should be changed (for either more fully descriptive names in our code or less in our tables), what do you think?

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java (hunk 9 above)
Comment (yardenvan) [reply src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java:203]: I changed this function according to your comment, however I do think it is still worthwhile having it as a function because it is used in several places and also for casting validation. Sounds okay?

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java (hunk 4 above)
Comment (itaitz) [reply src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: I'm not sure I fully follow... Do the 'assetTypeEnum' and the 'type' fields in Java represent the same field in the DB?

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java (hunk 9 above)
Comment (itaitz) [reply src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java:203]: LGTM

File: src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java (hunk 4 above)
Comment (yardenvan) [reply src/main/java/ResourceDiscovery/AssetObjects/AssetObject.java]: No, the "type" field is something returned from the assets (for example for an asset of Cloud SQL Instance it can be something like "mysql") and the assetTypeEnum does not have a field in the DB - its a constant used to decide which specific asset type should be created. 
Perhaps we can discuss this shortly tomorrow in our tri-weekly as I think it effects something a bit bigger then this in the design

File: src/main/java/ResourceDiscovery/README.md [hunk 11]
```diff
@@ -40,11 +40,38 @@
...
+    1. In the AssetTables enum class add a new enum by the following convention:
+    {SPECIFIC_ASSET_TYPE}_{GENERAL_ASSET_TYPE}_TABLE (for example: DISK_COMPUTE_TABLE)
+    Initialize it with the table name and a create table query (please use the provided helper
+    functions whenever possible).
+    1. In the ProjectMutationsList class you should add a new case statement in the addSpecificAssetMutation
+    function and set the new table fields into the relevant Mutation object.
 
+## Spanner DB Tables:
+### Adding a new asset table:
+[Please follow the steps as mentioned in the last section above.]()
```
Comment (itaitz) [inline src/main/java/ResourceDiscovery/README.md]: Why is this a link? (Seems like a markdown link...)

**Review by itaitz — APPROVED** [review]
Very nice! Just go over the remaining comments (if any) and submit
